lib/propolis/src/enlightenment/mod.rs

```diff
@@ -0,0 +1,105 @@
+//! directly with one another. This module distinguishes enlightenments from
+//! other kinds of virtual devices by the interfaces the guest uses to
+//! communicate with the host. For hypervisor enlightenments these include the
+//! following:
+//!
+//! 1. CPUID: The enlightenment stack injects synthetic CPUID values in a
+//!    well-known range of leaves (beginning with leaf 0x4000_0000) to advertise
+//!    its capabilities to guests.
+//! 2. Synthetic MSRs: The enlightenment stack intercepts RDMSR and WRMSR
+//!    instructions targeting MSRs in a well-known range (beginning with MSR ID
+//!    and interprets them according to its interface's definitions.
+//! 3. Direct sharing of guest physical memory: The guest can use MSRs to offer
+//!    to share its physical pages with the host, either to communicate directly
+//!    or for the host to overlay with larger blocks of information the guest
+//!    may wish to read.
+//! 4. Special VM exits: Both Intel's VMX and AMD's SVM provide special opcodes
+//!    (VMCALL and VMMCALL, respectively) that trigger a VM exit with a unique
+//!    exit code. The hypervisor can detect exits with this code, read the
+//!    guest's registers, and interpret them as parameters to a virtual function
+//!    call.
+//!
+//! Enlightenment stacks generally do not use port I/O or memory-mapped I/O
+//! to receive data from guests. This distinguishes them from other purely
+//! virtual devices (like virtio devices or the pvpanic device) that do not
+//! emulate any particular kind of physical hardware but nevertheless manifest
+//! themselves to the guest as attachments to a virtual bus.
+//!
+//! # This module
+//!
+//! This module defines traits that allow other Propolis components (notably
+//! vCPUs) to interact with an enlightenment stack. This module's submodules
+//! define various kinds of emulated hypervisor platforms and implement the
+//! enlightenments they supply.
+
+use std::sync::Arc;
+
+use cpuid_utils::CpuidSet;
+
+use crate::{
+    accessors::MemAccessor,
+    common::{Lifecycle, VcpuId},
+    msr::{MsrId, RdmsrOutcome, WrmsrOutcome},
+};
+
+pub mod bhyve;
+
+/// Functionality provided by all enlightenment interfaces.
+pub trait Enlightenment: Lifecycle + Send + Sync {
+    fn as_lifecycle(self: Arc<Self>) -> Arc<dyn Lifecycle>
+    where
+        Self: Sized,
+    {
+        self
+    }
+
+    /// Notifies this enlightenment stack that it has been adopted into a VM
+    /// with the supplied [`MemAccessor`] at the root of its memory accessor
+    /// hierarchy.
+    ///
+    /// Enlightenment stacks that want to access guest memory should call
+    /// [`MemAccessor::new_orphan`] when they are created, and then should call
+    /// [`MemAccessor::adopt`] from this function.
+    ///
+    /// Users of an enlightenment stack must guarantee that this function is
+    /// called only once per instance of that stack.
+    fn set_parent_mem_accessor(&self, parent: &MemAccessor);
```

Comment:
I would just name it something like `attach` (still passing in the `MemAccessor`), to make it a bit more generic, and in case we choose to wire up other bits (MMIO, etc) for other enlightments.